blocks/main_loop.py

```diff
@@ -156,16 +159,17 @@ def _check_finish_training(self):
         # the iteration the corresponding log record can be found only in
         # the previous row.
         if (self.log.current_row.training_finish_requested or
-                self.log.current_row.keyboard_interrupt_received or
-                self.log.previous_row.keyboard_interrupt_received):
+                self.log.current_row.interrupt_received or
+                self.log.previous_row.interrupt_received):
             raise TrainingFinish
 
-    def _handle_keyboard_interrupt(self, signal_number, frame):
+    def _handle_interrupt(self, signal_number, frame):
         # After receiving a first keyboard interrupt signal,
         # ignore all following ones.
```

Comment:
I am fine with your proposed SIGTERM behavior. As for waiting until the epoch's end: basically it means that a lot of users will have to always press it twice, because for many application epochs are very long. If we print something like "press another Ctrl + C to finish faster", this is also OK for me. 

And perhaps we should move printing related to handling interrupts from the `Printing` extension, because it is tightly related to interrupt handling logic which is currently simply hard-coded in the main loop. 

But well, this does not have be a part of this pull request, so if you want to have SIGTERM handled right away, you can merge it. Later I think we should consolidate all discussions on handling all sorts of interrupts and errors in one issue and decide what our policy is.